interpreter/resolver.go

```diff
@@ -0,0 +1,564 @@
+	//
+	// If the Attribute cannot be found, the return value must be a types.Err value.
+	ResolveRelative(interface{}, Activation, Attribute) interface{}
+}
+
+// NewResolver creates a Resolver from a type adapter.
+func NewResolver(adapter ref.TypeAdapter) Resolver {
+	return &defaultResolver{adapter: adapter}
+}
+
+// NewUnknownResolver creates a Resolver capable of inspecting a PartialActivation for the presence
+// of unknown values.
+func NewUnknownResolver(resolver Resolver) Resolver {
+	return &unknownResolver{
+		Resolver: resolver,
+	}
+}
+
+// ResolveStatus indicates the possible resolution outcomes for an Attribute.
+type ResolveStatus int
+
+const (
+	// FoundAttribute indicates that a top-level variable was found and that its path resolved
+	// to a value.
+	FoundAttribute ResolveStatus = 1 << iota
+
+	// NoSuchVariable indicates that the top-level variable was not provided in the Activation.
+	NoSuchVariable ResolveStatus = 1 << iota
+
+	// NoSuchAttribute indicates that a top-level variable was found, but the referenced Attribute
+	// could not be resolved.
+	NoSuchAttribute ResolveStatus = 1 << iota
+
+	// UnknownAttribute indicates that the Attribute path matched an unknown Attribute path.
+	UnknownAttribute ResolveStatus = 1 << iota
+)
+
+// ResolveListener receives an Attribute and the status of the Resolve call for each Attribute
+// provided to the Resolve call during the course of expression evaluation.
+type ResolveListener func(Attribute, ResolveStatus)
+
+// NewListeningResolver creates a Resolver that intercepts calls to Resolve and reports their
+// resolution status.
+func NewListeningResolver(resolver Resolver, listener ResolveListener) Resolver {
+	return &listeningResolver{
+		Resolver: resolver,
+		listener: listener,
+	}
+}
+
+// defaultResolver handles the resolution of attributes within simple Go native types.
+type defaultResolver struct {
+	adapter ref.TypeAdapter
+}
+
+// Resolve implements the Resolver interface method.
+func (res *defaultResolver) Resolve(vars Activation, attr Attribute) (interface{}, bool) {
+	varName := attr.Variable().Name()
+	attrPath := attr.Path()
+	obj, found := vars.Find(varName)
+	if found {
+		for _, elem := range attrPath {
+			obj = res.getElem(obj, elem.ToValue(vars))
+		}
+		return obj, true
+	}
+	return nil, false
+}
+
+// ResolveRelative implements the Resolver interface method.
+func (res *defaultResolver) ResolveRelative(
+	obj interface{},
+	vars Activation,
+	attr Attribute) interface{} {
+	for _, elem := range attr.Path() {
+		obj = res.getElem(obj, elem.ToValue(vars))
+	}
+	return obj
+}
+
+// getElem attempts to get the next 'elem' from the target 'obj' either by using native golang
+// method appropriate for the 'obj' type or by converting the obj to a ref.Val and using the CEL
+// functions.
+func (res *defaultResolver) getElem(obj interface{}, elem ref.Val) interface{} {
+	switch obj.(type) {
```

Comment:
Add a case for the error type, to make it clear that it's safe to continue to follow the elements even after one has failed.